public/app/features/dashboard-scene/serialization/DashboardSceneSerializer.ts

```diff
@@ -0,0 +1,106 @@
+import { config } from '@grafana/runtime';
+import { Dashboard } from '@grafana/schema';
+import { DashboardV2Spec } from '@grafana/schema/dist/esm/schema/dashboard/v2alpha0/dashboard.gen';
+import { SaveDashboardAsOptions } from 'app/features/dashboard/components/SaveDashboard/types';
+
+import { getRawDashboardChanges } from '../saving/getDashboardChanges';
+import { DashboardChangeInfo } from '../saving/shared';
+import { DashboardScene } from '../scene/DashboardScene';
+
+import { transformSceneToSaveModel } from './transformSceneToSaveModel';
+import { transformSceneToSaveModelSchemaV2 } from './transformSceneToSaveModelSchemaV2';
+
+export interface DashboardSceneSerializerLike<T> {
+  /**
+   * The save model which the dashboard scene was originally created from
+   */
+  initialSaveModel?: T;
+  getSaveModel: (s: DashboardScene) => T;
+  getSaveAsModel: (s: DashboardScene, options: SaveDashboardAsOptions) => T;
+  getDashboardChangesFromScene: (
+    scene: DashboardScene,
+    options: {
+      saveTimeRange?: boolean;
+      saveVariables?: boolean;
+      saveRefresh?: boolean;
+    }
+  ) => DashboardChangeInfo;
+}
+
+class V1DashboardSerializer implements DashboardSceneSerializerLike<Dashboard> {
+  initialSaveModel?: Dashboard;
+
+  getSaveModel(s: DashboardScene) {
+    return transformSceneToSaveModel(s);
+  }
+
+  getSaveAsModel(s: DashboardScene, options: SaveDashboardAsOptions) {
+    const saveModel = this.getSaveModel(s);
+
+    return {
+      ...saveModel,
+      id: null,
+      uid: '',
+      title: options.title || '',
+      description: options.description || '',
+      tags: options.isNew || options.copyTags ? saveModel.tags : [],
+    };
+  }
+
+  getDashboardChangesFromScene(
+    scene: DashboardScene,
+    options: { saveTimeRange?: boolean; saveVariables?: boolean; saveRefresh?: boolean }
+  ) {
+    const changedSaveModel = this.getSaveModel(scene);
+    const changeInfo = getRawDashboardChanges(
+      this.initialSaveModel!,
+      changedSaveModel,
+      options.saveTimeRange,
+      options.saveVariables,
+      options.saveRefresh
+    );
+
+    const hasFolderChanges = scene.getInitialState()?.meta.folderUid !== scene.state.meta.folderUid;
+
+    return {
+      ...changeInfo,
+      hasFolderChanges,
+      hasChanges: changeInfo.hasChanges || hasFolderChanges,
+    };
+  }
+}
+
+class V2DashboardSerializer implements DashboardSceneSerializerLike<DashboardV2Spec> {
+  initialSaveModel?: DashboardV2Spec;
+
+  getSaveModel(s: DashboardScene) {
+    return transformSceneToSaveModelSchemaV2(s) as DashboardV2Spec;
```

Comment:
This assertion can be avoided if we made `transformSceneToSaveModelSchemaV2` return just `DashboardV2Spec` and not `Partial<DashboardV2Spec> `. I'm wondering why we made it partial here because looking at v1, the `transformSceneToSaveModel` just returns `Dashboard` type.

Reply:
I think it's because at the time whe it was implemented, the type was very incomplete. I've changed this as of your suggestion.